### ISSUE TITLE
Review Column annotations

### ISSUE THREAD

**Issue by zero323** [zero323/pyspark-stubs#2]
Currently there is a number of problems with `pyspark.sql.column` annotations. Some are related to the Mypy behavior:

-  https://github.com/python/mypy/issues/2129 
- https://github.com/python/mypy/issues/2783

other, like `bitwise*` to the vague upstream semantics (should we allow `Any` if the only literal type acceptable on runtime is `int`?).

**Comment by harpaj** [comment]
@zero323,
The following valid snippet currently results in an error:

`df = df.withColumn("score", sum((F.col("sim"), F.col("weight"))))`
(note that sum is the stdlib sum, not the one from pyspark functions)

The error is `Argument 2 to "withColumn" of "DataFrame" has incompatible type "Union[Column, int]"; expected "Column"`

We actually do have a proper [annotation for `radd` on Column](https://github.com/zero323/pyspark-stubs/blob/master/third_party/3/pyspark/sql/column.pyi#L22), but it has a `type: ignore`. It sounds like this is connected to the python/mypy#2129 you mention above, but that one has been fixed. Do you think the `type: ignore` can be safely removed now?

**Comment by zero323** [comment]
@harpaj  Indeed, it looks like it should be safe to remove `type: ignore` now.

However I don't think that's really the source of the problem here.  With `ingore` this for example is valid:

```python
expr: Column = 1 + col("b")
```

It looks like the problem is more that we don't have dependent types here, and mypy cannot infer that `sum(NonEmptyIterable[Column])` is `Column`. Instead in consider both cases:

- iterable is empty and we default to `int`,
- iterable is non empty and we get `Column`

If you want `sum` to type check you should rather start with literal

```python
sum([F.col("sim"), F.col("weight")], lit(0))
```

**Comment by zero323** [comment]
Partially addressed by #194
